### PR TITLE
Instantiate execution plan capture callback via shim loader

### PR DESCRIPTION
**Pull request by abellina**
Closes https://github.com/NVIDIA/spark-rapids/issues/9103

This PR puts `ExecutionPlanCaptureCallback` under the shims so it can use shimmed classes such as `GpuFileSourceScanExec`. I was able to repro the failure seen in the MT CI job locally and this approach fixes it.

### CODE REVIEW

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
build

File: sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+{"spark": "322"}
+{"spark": "323"}
+{"spark": "324"}
+{"spark": "330"}
+{"spark": "330cdh"}
+{"spark": "330db"}
+{"spark": "331"}
+{"spark": "332"}
+{"spark": "333"}
+spark-rapids-shim-json-lines ***/
```
Comment (jlowe) [inline sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala:33]: This should not be in the shim directories.  There's no source code in this class that is specific to a version of Spark.  The only things that should be in the shim directories and using the shim-json-lines are files that have at least two different source code versions.  If the source code is identical across all Spark versions but compiles to different bytecodes, the build system will automatically handle that without needing to tag it for shimplify.

Comment (abellina) [reply sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala:33]: Ok will move that makes sense.

File: sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+{"spark": "321db"}
+{"spark": "322"}
+{"spark": "323"}
+{"spark": "324"}
+{"spark": "330"}
+{"spark": "330cdh"}
+{"spark": "330db"}
+{"spark": "331"}
+{"spark": "332"}
+{"spark": "333"}
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala]: Sorry I got confused and added shimplify directives and put this in the shims. I moved this per @jlowe's suggestion so we should not need this anymore

File: sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala (hunk 1 above)
Comment (abellina) [reply sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala]: This worked locally: https://github.com/NVIDIA/spark-rapids/pull/9113/commits/e17de2b43e019aee4fd2c027e51ddda3877fb409

**Comment by abellina** [comment]
build

**Review by jlowe — DISMISSED** [review]

**Comment by abellina** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.execution.{ExecSubqueryExpression, QueryExecution, ReusedSubqueryExec, SparkPlan}
+import org.apache.spark.sql.execution.adaptive.{AdaptiveSparkPlanExec, QueryStageExec}
+import org.apache.spark.sql.execution.exchange.ReusedExchangeExec
+
+
+/**
+ * Note that the name is prefixed with "Shimmed" such that wildcard rules
+ * under unshimmed-common-from-spark311.txt don't get confused and pick this class to be
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala:34]: Happy to file something. It seems we should likely pick a standard prefix for something that should be unshimmed? Maybe a standard package would help here…

[the remark this answers is not in the document]
Comment (gerashegalov) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala:34]: The issue with the package is that we may need a specific package to get access to package-private scope. Looks like there is definitely enough content for a separate issue

[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/ShimmedExecutionPlanCaptureCallbackImpl.scala:34]: I am wondering if https://github.com/NVIDIA/spark-rapids/issues/6565 will resolve this for us, so we don't have to worry about the naming anymore.

**Comment by abellina** [comment]
This should be ready to go in. If folks here don't mind to take a look @jlowe @gerashegalov @andygrove

**Review by jlowe — APPROVED** [review]

**Review by gerashegalov — APPROVED** [review]
LGTM

**Pull request merged**